src/vm/runners/cairo_runner.rs

```diff
@@ -593,11 +594,26 @@ impl CairoRunner {
             .map_err(VirtualMachineError::TracerError)?;
         vm.end_run(&self.exec_scopes)?;
 
-        if !disable_finalize_all {
-            vm.segments.compute_effective_sizes(&vm.memory);
-            self.run_ended = true;
+        if disable_finalize_all {
+            return Ok(());
+        }
+
+        vm.segments.compute_effective_sizes(&vm.memory);
+        if self.proof_mode && !disable_trace_padding {
+            self.run_until_next_power_of_2(vm, hint_processor)?;
+            loop {
+                if let Err(e) = self.check_used_cells(vm) {
+                    if e == MemoryError::InsufficientAllocatedCells.into() {
+                        break;
+                    }
+                }
+
+                self.run_for_steps(1, vm, hint_processor)?;
+                self.run_until_next_power_of_2(vm, hint_processor)?;
+            }
         }
```

Comment:
Is this behavior correct?, in the original vm the loop ends when check_used_cells returns True (our counterpart would be Ok()), but we end the loop when we get an InsufficientAllocatedCells error.

Reply:
Yes, should be correct since `check_used_cells()` does catch `InsufficientAllocatedCells()` to return `False` which then breaks this loop.

Maybe it'd be better to move that special case into `check_used_cells()` as in Python?

Reply:
What im worried about is that in python the loop runs while check_used_cells() returns False, and ends when it returns True `while not self.check_used_cells():`, but we are breaking the loop when it returns False (aka InsufficientAllocatedCells)

Reply:
Oh, you're right. That makes more sense. Thanks.

Reply:
Fixed that and another bug where I forgot to return when an error different from that special case is returned.